Orange/data/table/base.py

```diff
@@ -0,0 +1,1271 @@
+import os
+import zlib
+from collections import Sequence
+from threading import Lock
+from numbers import Number
+
+from itertools import chain
+import numpy as np
+import scipy.sparse as sp
+import pandas as pd
+import pandas.core.internals
+
+from Orange.data import Domain, StringVariable, ContinuousVariable, DiscreteVariable, Variable, TimeVariable
+from Orange.util import flatten, deprecated
+
+
+# noinspection PyPep8Naming
+class TableBase:
+    """An abstract base class for data storage structures in Orange."""
+    KNOWN_PANDAS_KWARGS = {}
+
+    # the default name for weights columns
+    # subclasses may override this to rename the column
+    _WEIGHTS_COLUMN = "__weights__"
+
+    # a collection of all known internal column names
+    # use case: Corpus adds additional columns and we need to know
+    # of all available names
+    _INTERNAL_COLUMN_NAMES = [_WEIGHTS_COLUMN]
+
+    # a counter for indexing rows, important for deterministically selecting rows
+    # and keeping pandas indices sane
+    _next_instance_id = 0
+    _next_instance_lock = Lock()
+
+    conversion_cache = None
+
+    # custom properties, preserved through pandas manipulations
+    _metadata = ['name',
+                 'domain',
+                 'attributes',
+                 '__file__']
+
+    def __new__(cls, *args, **kwargs):
+        """Create a new Table, the exact result type depends on the data passed.
+
+        If called without arguments, create and initialize a blank Table, otherwise
+        intelligently call one of the TableBase.from_X functions, depending on the arguments.
+        Also passes through pandas.DataFrame constructor keyword arguments.
+        Does not pass positional arguments through to pandas, unless called from pandas internals.
+
+        Parameters
+        ----------
+        args
+            One of:
+             - (str): create a Table from a file or URL
+             - (TableBase): initialize from another TableBase,
+             - (np.ndarray | scipy.sparse): a dense or sparse matrix, resulting
+               Table data type depends on the type passed, infer a domain
+                - Up to four may be passed to fit into X/Y/metas/weights respectively.
+             - (Domain, np.ndarray | scipy.sparse): Same as above, without domain inference
+             - (Domain, TableBase): convert another TableBase object to the passed domain
+             - (Domain, list): create a Table based on a list of rows.
+             - (Domain, pd.DataFrame): create a table based on an existing pandas DataFrame and Domain.
+        kwargs
+            Used exclusively for compatibility with pandas.
+
+        Returns
+        -------
+        A new Table or SparseTable, depending on the arguments passed and the calling class.
+        """
+        # Needed because we have two construction paths: Table() or Table.from_X,
```

Comment:
I now moved it under the pandas compatibility section, it was referring to that.